### PR TITLE
[netatmo] Consolidate OAuth2 by using core implementation and storage

### LINKED ISSUE #14755: [netatmo] Consolidate OAuth2 handling
Recently Netatmo announced changes in their OAuth2 implementation to make it RFC compliant: #14546

On our side #14548 aligned the openHAB implementation with those changes, but with a custom OAuth2 implementation.

Since I have recently had a look at OAuth2 in context of #14745, I think it would be worth streamlining the OAuth2 implementation by using the available openHAB core classes.

According to the [Netatmo documentation](https://dev.netatmo.com/apidocumentation/oauth), the API should be fully [RFC 6749](https://www.rfc-editor.org/rfc/rfc6749) compliant, thus it should work.

In case any obstacles are found, this should be possible to solve, even it core changes would be needed. RFC violations should preferably be handled directly in the binding, but the core classes should make it possible to extend the functionality to make this possible.

I'm aware of https://github.com/openhab/openhab-addons/pull/14548#issuecomment-1459744332, so this issue also serves as a placeholder for documenting those potential problems, and coming up with solutions.

If I understand the problem correctly, the grant response looks like this:

```json
{
	"access_token": "51c55d0418xxxxxxxxfa",
	"refresh_token": "51c55xxxxxxxxeb3b",
	"scope": [
		"read_station",
		"read_homecoach",
		"read_presence",
		"access_presence",
		"write_camera",
		"access_camera",
		"read_thermostat",
		"write_thermostat",
		"read_smokedetector",
		"read_camera"
	],
	"expires_in": 10800,
	"expire_in": 10800
}
```

which is expected to be:

```json
{
	"access_token": "51c55d0418xxxxxxxxfa",
	"refresh_token": "51c55xxxxxxxxeb3b",
	"scope": "read_station read_homecoach read_presence access_presence write_camera access_camera read_thermostat write_thermostat read_smokedetector read_camera",
	"expires_in": 10800,
	"expire_in": 10800
}
```

[Netatmo reply](https://forum.netatmo.com/viewtopic.php?t=19293) March 13rd 2023: _"It's not something planned for the moment, but teams took the point and will think about it"_

It seems like openhab/openhab-core#1891 should make it possible to work around.

Perhaps [importAccessTokenResponse](https://www.openhab.org/javadoc/latest/org/openhab/core/auth/client/oauth2/oauthclientservice#importAccessTokenResponse(org.openhab.core.auth.client.oauth2.AccessTokenResponse)) could be a second option.

@clinique, @lolodomo - FYI.

### PR DESCRIPTION
**Pull request by jlaur**
This replaces the custom OAuth2 authorization requests and storage by the implementation in [org.openhab.core.auth.client.oauth2](https://www.openhab.org/javadoc/latest/org/openhab/core/auth/client/oauth2/package-summary.html).

This has some benefits:
- Only one implementation to maintain and with full feature set.
- Shared **jsondb** storage for all implementations, which will be included in backups.

Additionally, the behavior of multiple fighting clients have changed. Previously, the bridge would get status "OFFLINE (COMMUNICATION_ERROR): Bad Request" when reinitializing the bridge handler after authorizing another client. I noticed this on my OH 3.4 production system while testing the authorization flow on a 4.0 system. Now, the reauthorization flow will be correctly started.

After completing the authorization flow successfully, the servlet will now be disposed.

Depends on openhab/openhab-core#3537
Fixes #14755

### CODE REVIEW

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/AuthenticationApi.java
Comment (clinique) [inline bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/AuthenticationApi.java:47]: `scheduler` could be removed from the parameter list also.

**Review by clinique — APPROVED** [review]
Mmmm !  I like that.

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/AuthenticationApi.java
Comment (jlaur) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/AuthenticationApi.java]: Good catch!

**Comment by lolodomo** [comment]
Thank you for your review @clinique , I was about to ask you. 
I will install the last snapshot this morning in order to test that change. 
And I plan to review the PR today. 
Objective: having that change in milestone 2.

**Comment by lolodomo** [comment]
Snapshot 3413 installed and PR compiled and installed. I was able to authenticate and all my things are again ONLINE. Great.

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:137]: Please move this line below, at least after your 2 checks of valid client id and client secret. Or just before your call to scheduler.execute.
This will avoid a doubled thing state update.

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/dto/NetatmoAccessTokenResponse.java
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/dto/NetatmoAccessTokenResponse.java:98]: Please add something or remove this line.

**Review by lolodomo — CHANGES_REQUESTED** [review]
Great job @jlaur 
I have only 2 minor comments.

**Comment by lolodomo** [comment]
@jlaur I just took a look to the content of my file StorageHandler.For.OAuthClientService.json.
I use now 3 bindings using OAuth2 authentications: ecowatt, netatmo, homeconnect.
For each of them, I have 3 blocks: AccessTokenResponse, LastUsed and ServiceConfiguration.
The file also starts with that:
```
INDEX_HANDLES": {
    "class": "java.lang.String",
    "value": "[\n  \"netatmo:account:compte\",\n  \"ecowatt:signals:monecowatt\"\n]"
  },
```
Is it normal that I do not have also "homeconnect:api_bridge:bridge" ?
Could it be a bug in the core framework ?

**Comment by lolodomo** [comment]
The access token has apparently a validity of 3 hours. So to be sure that everything is working well, we have to wait for 3 hours to be sure that the token is refreshed properly.
But I assumed you (@jlaur and @clinique) already checked that ?

**Review by lolodomo — APPROVED** [review]
LGTM, big thank you for this contribution.

**Comment by jlaur** [comment]
> The access token has apparently a validity of 3 hours. So to be sure that everything is working well, we have to wait for 3 hours to be sure that the token is refreshed properly.

Let's wait to be sure.

> But I assumed you (@jlaur and @clinique) already checked that ?

I believe so, but I also have collisions between my production system and test system, so frankly I don't remember if I actually tried disconnecting production for three hours to be sure not to have any interference during token renewal.

**Comment by jlaur** [comment]
> Is it normal that I do not have also "homeconnect:api_bridge:bridge" ?
> Could it be a bug in the core framework ?

I would have to check, but I noticed the same. Perhaps this is handles currently opened, but I would assume you should then still see all three.

**Comment by lolodomo** [comment]
> I would have to check, but I noticed the same. Perhaps this is handles currently opened, but I would assume you should then still see all three.

My device is OFFLINE.
Let me switch it ON.

**Comment by lolodomo** [comment]
> I don't remember if I actually tried disconnecting production for three hours to be sure not to have any interference during token renewal.

I was not thinking about that special case but rather the normal case that OH will have to refresh the token after 3 hours.
I mean just checking that the binding is still retrieving data after few hours.

**Comment by jlaur** [comment]
> I was not thinking about that special case but rather the normal case that OH will have to refresh the token after 3 hours.
> I mean just checking that the binding is still retrieving data after few hours.

I understand, I'm just in doubt if I actually fully tested the normal scenario, so it would be much appreciated if you could leave it running for those hours.

**Comment by lolodomo** [comment]
> I understand, I'm just in doubt if I actually fully tested the normal scenario, so it would be much appreciated if you could leave it running for those hours.

No problem, I just have to wait few hours to notice that the binding is still working.
I will have to restart OH to be sure.

**Comment by lolodomo** [comment]
Bad news, it looks like my data in OH are now outdated (not the one on the netatmo app). The things are still ONLINE.

**Comment by lolodomo** [comment]
Here are logs, the token is apparently not refreshed after expiration:
```
2023-04-15 15:13:55.267 [WARN ] [al.handler.capability.HomeCapability] - Error getting Home informations : Rest call failed: statusCode=ACCESS_TOKEN_EXPIRED, message=Access token expired
2023-04-15 15:13:55.304 [WARN ] [andler.capability.SecurityCapability] - Error retrieving last events for home 'xxx' : Rest call failed: statusCode=ACCESS_TOKEN_EXPIRED, message=Access token expired
2023-04-15 15:13:55.310 [DEBUG] [handler.capability.RefreshCapability] - Module refreshed, next one in 150 s
2023-04-15 15:16:25.474 [WARN ] [al.handler.capability.HomeCapability] - Error getting Home informations : Rest call failed: statusCode=ACCESS_TOKEN_EXPIRED, message=Access token expired
2023-04-15 15:16:25.513 [WARN ] [andler.capability.SecurityCapability] - Error retrieving last events for home 'xxx' : Rest call failed: statusCode=ACCESS_TOKEN_EXPIRED, message=Access token expired
2023-04-15 15:16:25.518 [DEBUG] [handler.capability.RefreshCapability] - Module refreshed, next one in 150 s
...
2023-04-15 15:19:45.086 [WARN ] [handler.capability.WeatherCapability] - Error retrieving weather data 'yyyy' : Rest call failed: statusCode=ACCESS_TOKEN_EXPIRED, message=Access token expired
2023-04-15 15:19:45.089 [DEBUG] [handler.capability.RefreshCapability] - Module refreshed, next one in 120 s
2023-04-15 15:21:25.953 [DEBUG] [handler.capability.RefreshCapability] - Module refreshed, next one in 150 s
2023-04-15 15:21:45.129 [WARN ] [handler.capability.WeatherCapability] - Error retrieving weather data 'yyyy' : Rest call failed: statusCode=ACCESS_TOKEN_EXPIRED, message=Access token expired
2023-04-15 15:21:45.132 [DEBUG] [handler.capability.RefreshCapability] - Module refreshed, next one in 120 s
2023-04-15 15:23:45.294 [WARN ] [handler.capability.WeatherCapability] - Error retrieving weather data 'yyyy' : Rest call failed: statusCode=ACCESS_TOKEN_EXPIRED, message=Access token expired
2023-04-15 15:23:45.297 [DEBUG] [handler.capability.RefreshCapability] - Module refreshed, next one in 120 s
...
```

**Comment by lolodomo** [comment]
Additionally, what is strange is that the binding doesn't care about the token expiration and the things are still all ONLINE. 

If I restart the binding, it looks like I got fresh data, probably with a new token. I guess it will work 3 hours.

**Comment by lolodomo** [comment]
Off-topic: I checked how many bindings are already using the OAuth2 client from the core framework: 17.
On these 17 bindings, only 9 are calling `ungetOAuthService`; 8 do not and this is probably missing: hydrawise, livisismarthome, mercedesmen, mielecloud, myq, nest, googlestt and googletts.
On these 17 bindings, only ecobee and mielecloud are calling `deleteServiceAndAccessToken`; all other (15) don't and this is probably missing.

**Comment by jlaur** [comment]
> Bad news, it looks like my data in OH are now outdated (not the one on the netatmo app). The things are still ONLINE.

Thanks for testing it. I have not been testing this main scenario correctly, so just had a look at the code again, and it's a pretty obvious bug. I'll fix it and retest properly.

**Comment by jlaur** [comment]
> Thanks for testing it. I have not been testing this main scenario correctly, so just had a look at the code again, and it's a pretty obvious bug. I'll fix it and retest properly.

As suspected, it was a pretty stupid bug. I have fixed it, but will now need to wait three hours for confirmation. Additionally, I would like to refactor this. I don't like the coupling of `ApiBridgeHandler` and `AuthenticationApi`. Probably `AuthenticationApi` is not needed anymore. It should not be possible to ask for the authorization header without going through the logic responsible for refreshing the tokens.

**Comment by lolodomo** [comment]
Not good, even worst, it now tries to connect several times per second, in something that seems to be an infinite loop.

**Comment by lolodomo** [comment]
So I guess it is too late for this PR for the milestone 2.

**Comment by jlaur** [comment]
> Not good, even worst, it now tries to connect several times per second, in something that seems to be an infinite loop.

I cannot reproduce this. Do you have some logs to share?

With trace logging enabled:
```shell
openhab> log:set trace org.openhab.binding.netatmo
openhab> log:set trace org.openhab.core.auth.oauth2client
```

I now see this periodically:

```
2023-04-15 20:49:02.619 [TRACE] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getstationsdata?device_id=xxx&get_favorites=false 
2023-04-15 20:49:02.620 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - Connecting to Netatmo API.
2023-04-15 20:49:02.621 [TRACE] [mo.internal.handler.ApiBridgeHandler] -  -with headers : Accept-Encoding: gzip, User-Agent: Jetty/9.4.50.v20221201, Authorization: Bearer yyy
2023-04-15 20:49:02.836 [TRACE] [mo.internal.handler.ApiBridgeHandler] -  -returned : code [200 OK] body [...]
2023-04-15 20:49:02.839 [DEBUG] [handler.capability.RefreshCapability] - Module refreshed, next one in 603 s
2023-04-15 20:59:05.842 [TRACE] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getstationsdata?device_id=xxx&get_favorites=false 
2023-04-15 20:59:05.846 [DEBUG] [lient.internal.OAuthStoreHandlerImpl] - Decrypting token: AccessTokenResponse [accessToken=zzz, tokenType=null, expiresIn=10800, refreshToken=qqq, scope=read_homecoach read_station read_thermostat write_thermostat read_doorbell access_camera read_smokedetector write_doorbell access_doorbell write_camera read_presence write_presence read_carbonmonoxidedetector read_camera access_presence, state=null, createdOn=2023-04-15T17:59:34.623108500Z]
2023-04-15 20:59:05.847 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - Connecting to Netatmo API.
2023-04-15 20:59:05.848 [TRACE] [mo.internal.handler.ApiBridgeHandler] -  -with headers : Accept-Encoding: gzip, User-Agent: Jetty/9.4.50.v20221201, Authorization: Bearer yyy
2023-04-15 20:59:06.049 [TRACE] [mo.internal.handler.ApiBridgeHandler] -  -returned : code [200 OK] body [...]
2023-04-15 20:59:06.052 [DEBUG] [handler.capability.RefreshCapability] - Module refreshed, next one in 605 s
```

I'm now monitoring **StorageHandler.For.OAuthClientService.json**, but still a few hours to go.

**Comment by jlaur** [comment]
> So I guess it will be too late for this PR for the milestone 2.

I agree, at this point it would be too risky for M2. Let's get it right and properly tested. The `WebhookServlet` also needs to be handled a bit differently/separately to avoid recreation on each request.

**Comment by jlaur** [comment]
@lolodomo - finally my tests were concluded after three hours. It's now working correctly. Relevant logs:
```
2023-04-15 19:59:06.966 [INFO ] [etatmo.internal.servlet.GrantServlet] - Registered Netatmo servlet at '/netatmo/connect/xxx'
2023-04-15 19:59:34.631 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:account:b61fd88486' changed from OFFLINE (CONFIGURATION_ERROR): Configuration incomplete, please grant the binding to Netatmo Connect. to ONLINE
[...]
2023-04-15 23:00:12.504 [TRACE] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getstationsdata?device_id=zzz&get_favorites=false 
2023-04-15 23:00:12.505 [DEBUG] [lient.internal.OAuthStoreHandlerImpl] - Decrypting token: AccessTokenResponse [accessToken=xxx, tokenType=null, expiresIn=10800, refreshToken=aaa, scope=read_homecoach read_station read_thermostat write_thermostat read_doorbell access_camera read_smokedetector write_doorbell access_doorbell write_camera read_presence write_presence read_carbonmonoxidedetector read_camera access_presence, state=null, createdOn=2023-04-15T17:59:34.623108500Z]
2023-04-15 23:00:12.506 [DEBUG] [lient.internal.OAuthStoreHandlerImpl] - Decrypting token: AccessTokenResponse [accessToken=xxx, tokenType=null, expiresIn=10800, refreshToken=aaa, scope=read_homecoach read_station read_thermostat write_thermostat read_doorbell access_camera read_smokedetector write_doorbell access_doorbell write_camera read_presence write_presence read_carbonmonoxidedetector read_camera access_presence, state=null, createdOn=2023-04-15T17:59:34.623108500Z]
2023-04-15 23:00:12.515 [DEBUG] [oauth2client.internal.OAuthConnector] - Oauth request parameter grant_type, value refresh_token
2023-04-15 23:00:12.515 [DEBUG] [oauth2client.internal.OAuthConnector] - Oauth request parameter refresh_token, value ddd
2023-04-15 23:00:12.515 [DEBUG] [oauth2client.internal.OAuthConnector] - Oauth request parameter scope, value read_homecoach read_station read_thermostat write_thermostat read_doorbell access_camera read_smokedetector write_doorbell access_doorbell write_camera read_presence write_presence read_carbonmonoxidedetector read_camera access_presence
2023-04-15 23:00:12.516 [DEBUG] [oauth2client.internal.OAuthConnector] - Setting authentication for clientId ccc. Using basic auth false
2023-04-15 23:00:18.711 [DEBUG] [oauth2client.internal.OAuthConnector] - grant type refresh_token to URL https://api.netatmo.com/oauth2/token success
2023-04-15 23:00:18.716 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - Connecting to Netatmo API.
2023-04-15 23:00:18.717 [TRACE] [mo.internal.handler.ApiBridgeHandler] -  -with headers : Accept-Encoding: gzip, User-Agent: Jetty/9.4.50.v20221201, Authorization: Bearer qqq 
2023-04-15 23:00:20.608 [TRACE] [mo.internal.handler.ApiBridgeHandler] -  -returned : code [200 OK] body [...]
2023-04-15 23:00:20.610 [TRACE] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getstationsdata?device_id=zzz&get_favorites=false 
2023-04-15 23:00:20.610 [DEBUG] [lient.internal.OAuthStoreHandlerImpl] - Decrypting token: AccessTokenResponse [accessToken=yyy, tokenType=null, expiresIn=10800, refreshToken=bbb, scope=read_homecoach read_station read_thermostat write_thermostat read_doorbell access_camera read_smokedetector write_doorbell access_doorbell write_camera read_presence write_presence read_carbonmonoxidedetector read_camera access_presence, state=null, createdOn=2023-04-15T21:00:18.711607200Z]
2023-04-15 23:00:20.611 [DEBUG] [handler.capability.RefreshCapability] - Module refreshed, next one in 595 s
[...]
2023-04-15 23:06:24.008 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:account:b61fd88486' changed from ONLINE to UNINITIALIZED
2023-04-15 23:06:24.013 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:account:b61fd88486' changed from UNINITIALIZED to UNINITIALIZED (DISABLED)
2023-04-15 23:06:27.131 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:account:b61fd88486' changed from UNINITIALIZED (DISABLED) to INITIALIZING
2023-04-15 23:06:27.137 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:account:b61fd88486' changed from INITIALIZING to UNKNOWN
2023-04-15 23:06:27.138 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:account:b61fd88486' changed from UNKNOWN to ONLINE
```

**Comment by jlaur** [comment]
> I agree, at this point it would be too risky for M2. Let's get it right and properly tested. The `WebhookServlet` also needs to be handled a bit differently/separately to avoid recreation on each request.

@lolodomo, @clinique - I fixed the `WebhookServlet` logic in the last commit, and my own tests of the main road scenario were successful. After token expiration, it's correctly renewed and stored in jsondb, and will also go back online after disabling/enabling the thing.

I don't know if you'll have any time today, but if would also find it working, it might be back on the table for M2? @lolodomo - before any such considerations, we'll definitely need to know more about that infinite loop you mentioned yesterday.

Still, I would not rush it, but if it works for you also, it could be included. The advantage would be skipping the intermediate step with the custom implementation and file storage. Only tomorrow will we know for sure if it works fully as intended - which I'm pretty sure the core OAuth2 implementation will.

**Comment by lolodomo** [comment]
> I don't know if you'll have any time today, but if would also find it working, it might be back on the table for M2? @lolodomo - before any such considerations, we'll definitely need to know more about that infinite loop you mentioned yesterday.

I will not have much time today, sorry. 
Maybe @clinique can already test with his netatmo setup?

**Comment by clinique** [comment]
@lolodomo, won't probably have time today.

**Comment by lolodomo** [comment]
I just tried against but this time I restartted openHAB. I did not encounter the problem I noticed yesterday.

**Comment by lolodomo** [comment]
My values in OH are still in line with the netatmo app after 8 hours.

**Comment by lolodomo** [comment]
Seems to work perfectly for me, I plan to merge this PR this evening after reviewing the last commits.

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:165]: Something that might be strange. You are not considering the result of authenticate and just after you start the webhook servlet and set the thing status to ONLINE.
But if `authenticate` leads to the call of `startAuthorizationFlow`, I have the feeling you should not do what is just after this line ?

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:173]: You should first dispise the servlet before creating a new one?

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java
Comment (jlaur) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:165]: I agree this can be refactored further.

I believe the current logic corresponds to the previous behavior, so as a consequence of keeping the scope to replacing the OAuth2 implementation, this has not yet been addressed.

Previously we would read refresh token from file, and if not blank, the logic from line 164 (`case COMPLETED`) in the old code on the left side would be triggered.

I already had some additional refactoring on my TODO for a follow-up PR, but I can also enhance it further in this PR.

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java
Comment (jlaur) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:173]: Same situation as above, this is a lack of further enhancements, not something introduced.

Comment (lolodomo) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:173]: It looks easy to fix. Could be done in this PR.

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java
Comment (lolodomo) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:165]: I just do not understand how we can be in status configuration error with the current code. 
Strangely I saw this status before authenticating. 
There is certainly something obvious I do not see because my feeling is that the current code should always lead to an ONLINE status... except if there is an uncaught exception preventing to reach the end of this method.

Comment (jlaur) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:165]: You are right:
```
2023-04-21 15:14:58.004 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:account:b61fd88486' changed from UNINITIALIZED (DISABLED) to INITIALIZING
2023-04-21 15:14:58.013 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:account:b61fd88486' changed from INITIALIZING to UNKNOWN
2023-04-21 15:14:58.188 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:account:b61fd88486' changed from UNKNOWN to OFFLINE (CONFIGURATION_ERROR): Configuration incomplete, please grant the binding to Netatmo Connect.
2023-04-21 15:14:58.188 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:account:b61fd88486' changed from OFFLINE (CONFIGURATION_ERROR): Configuration incomplete, please grant the binding to Netatmo Connect. to ONLINE
```

**Review by lolodomo — APPROVED** [review]
LGTM

**Comment by lolodomo** [comment]
I hope you have done few tests with the last commit.
I merge without testing again myself.

**Pull request merged**
